bt_ddos_shield/miner_shield.py

```diff
@@ -1,74 +1,481 @@
+        self.options = options
+
+        self.worker_thread = None
+        self.task_queue = Queue()
+        self.run = False
+        self.finishing = False
+        self.ticker = threading.Event()
+        self.ticker_thread = None
 
     def enable(self):
         """
-        Enable shield. It asynchronously starts the shield, which consists of such steps:
+        Enable shield. It starts worker thread, which will do such steps if run for the first time:
         1. Fetch validators keys.
         2. Creates addresses for all validators.
         3. Save manifest file.
         4. Publish link to manifest file to blockchain.
         5. Eventually close public access to original IP after some time.
 
-        It puts events to event_manager after each step. Current state is managed by state_manager. If shielding
-        process had been interrupted it is continued from the last step.
+        It puts events to event_manager after each finished operation. Current state is managed by state_manager.
+        If any error occurs it is retried forever until shield is disabled.
 
-        When shield is running, changing validators set triggers shield reconfiguration.
+        When shield is running, user can schedule tasks to be processed by worker.
         """
-        pass
+        if self.worker_thread is not None:
+            # already started
+            return
+
+        self.finishing = False
+        self.run = True
+        self._add_task(MinerShieldInitializeTask())
+        self.worker_thread = threading.Thread(target=self._worker_function)
+        self.worker_thread.start()
+
+    def disable(self):
+        """
+        Disable shield. It stops worker thread after finishing current task. Function blocks until worker is stopped.
+        """
+        self._add_task(MinerShieldDisableTask())
+
+        self.finishing = True
+        self.ticker.set()
+
+        if self.worker_thread is not None:
+            self.worker_thread.join()
+            self.worker_thread = None
+
+        self.task_queue = Queue()  # clear task queue
+
+        if self.ticker_thread is not None:
+            self.ticker_thread.join()
+            self.ticker_thread = None
 
     def ban_validator(self, validator_hotkey: Hotkey):
         """
-        Ban a validator by its hotkey. Function blocks execution until manifest file is updated and info about file
-        is published to Bittensor.
+        Ban a validator by its hotkey. Task will be executed by worker. It will update manifest file and publish info
+        about new file version to blockchain.
+        """
+        self._add_task(MinerShieldBanValidatorTask(validator_hotkey))
+
+    def _add_task(self, task: 'AbstractMinerShieldTask'):
+        """
+        Add task to task queue. It will be handled by _worker_function.
+        """
+        if not self.run:
+            raise MinerShieldDisabledException()
+
+        self.task_queue.put(task)
+
+    def _worker_function(self):
```

Comment:
for example `def _add_task(self, task: 'AbstractMinerShieldTask'):` could mostly be replaced with `await task`

Reply:
I'm not sure if it can be done with asyncio tasks, because I don't wait here for the result and task is retried forever by worker until shield gets stopped (this behaviour is by design). I will learn more about asyncio and then think if it can be changed, because my knowledge about asyncio is very small now. I wrote it in my notes for future upgrade but as for now I think this can be left as it is.

Reply:
I read something, thought about it, used AI and probably this can be refactored. But when current behavior is to be maintained, using asyncio won't change too much - I can use asyncio.to_thread to run _worker_function, use asyncio.Queue() instead of just Queue and make rest of needed changes, but logic and code structure will looks nearly the same (if I'm thinking correctly). So my question is, if it is worth time to change it, if it already works? I don't know the answer... Next time, when doing similar job, I'll try to use asyncio, but should I spend time for changing it here?